pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -432,8 +432,16 @@ func (r *ReconcileBareMetalHost) actionDeleting(prov provisioner.Provisioner, in
 		info.log.Info("ready to be deleted")
 		return deleteComplete{}
 	}
-
-	provResult, err := prov.Delete()
+	// try to power off host before delete
+	provResult, err := prov.PowerOff()
```

Comment:
What happens here if the BMC credentials are invalid? Do we sit in a reconcile loop forever until they're fixed, and then the host is deleted?

Reply:
If we were able to provision the host earlier, does that not imply that we have the correct BMC credentials?

Reply:
They could have been changed out from under us, on the device itself. That would mean Ironic would report an error trying to turn off the power, and here we now always treat that error as something that needs a retry. I wonder if we want to limit the number of retries and at some point just ignore the error? Or maybe move the power off operation in the Delete() method of the provisioner, where it can behave more intelligently based on the specific error .

Reply:
We could call `ValidateManagementAccess()` and only try the power off if it succeeds. There's a risk that a temporary failure could prevent the power being turned off, but I think we're mainly doing this to make testing easier anyway so maybe it's not a big deal.

Reply:
That might work.

Reply:
I am concerned about any unintended consequences from https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/ironic.go#L384 being called multiple times in the delete path via ValidateManagementAccess().
If https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/ironic.go#L228 is the only one that is actually verifying BMC credentials, the why not call just that from PowerOff() before attempting a poweroff and give a better error message about what exactly failed.

Reply:
To make the code more efficient, I would like to propose this:
1. Call poweroff()
2. if previous call returns error, then check for credentials
3. If credentials are bad, ignore all poweroff() errors, report issue with bad credentials, proceed to Delete()
4. If credentials are good [and poweroff() still failed], we could call actionError() [which will result in retry]
5. Delete()

Assuming that credentials changing underneath us is a fairly infrequent occurance, this way we make the expensive call to the BMC to verify credentials only in the case of an error.

Reply:
Checking the credentials doesn't necessarily happen in a single call, it's something that could occur over the course of several requeues. Despite that, this might still work though because PowerOff() doesn't change the state.
However, we'd still have to deal with the case where we get up to calling Delete(), and it succeeds but we fail when trying to remove the finaliser. If that happens, on the next reconcile we will call PowerOff() and it will fail to load the node from Ironic, and from that point we will be in an infinite loop.

I think we need to:
* Set an ErrorMessage in PowerOff() when a previous attempt to power off the host has failed.
* Ignore both the Dirty flag and the ErrorMessage if we get an ErrorMessage from PowerOff().
* Ignore any err returned from PowerOff().

Reply:
Reading this comment https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/ironic.go#L1117-L1120 has spawned a related question. Do we have to make sure the node is not externally provisioned before powering it off?

Reply:
Excellent question, @sadasu. What behavior would you expect?

Reply:
We should not be powering off that node. But, what I am unclear about is whether the responsibility for checking if the node is externally provisioned lies with the  baremetal controller with actionDeleting() or with the powerOff() logic in the provisioner (Ironic). Since the provisioner is already aware if a node is externally provisioned, I think that it should be the provisioner that makes sure it doesn't power off a node that it did not provision.

Reply:
Another good question. I think the context matters for the answer to that one.

The provisioner does know whether a host is externally provisioned. However, it doesn't know inside PowerOff() what the context for that request is. For example, it doesn't know if the host is being deprovisioned or put into maintenance mode. I could see us wanting to support maintenance mode for externally provisioned hosts, so I think we want to make the power control decision outside of the provisioner.

Reply:
Great catch Sandhya.
I agree with Doug, IMHO it should be the controller that decides not to power off an externally-provisioned node. If we had another provisioner implementation we wouldn't want it to have different behaviour, so it shouldn't be left up to the provisioner. And to the extent possible the provisioner methods shouldn't behave differently depending on the host state; it's the job of the controller to manage state.

Reply:
I think this is done too.